rfcs/NNNN-hot-elements.md

```diff
@@ -0,0 +1,77 @@
+
+### Goals
+- Instantiating a new custom element, post HMR, should use the newest version of the constructor.
+- Provide an API for custom element classes so that they can add additional handling of HMR. For example, in LitElement we can patch and rerender existing elements in the page.
+  - This API should be implementable by a base class, so that in many cases no changes are needed to an element for it to achieve optimal HMR.
+- Support @web/dev-server-hmr specifically and the [esm-hmr](https://github.com/FredKSchott/esm-hmr) protocol more generally.
+
+### Non-Goals
+- Support all HMR servers and protocols, at least initially.
+- Work perfectly. Need to set expectations that, due to fundamental JS semantics, HMR can't and won't always be perfect.
+
+## Motivation
+
+HMR can be a boon to dev productivity by reducing the latency of the edit/refresh loop, particularly for larger apps.
+
+## Detailed Design
+
+We'll start by building off of [redefine-custom-elements](https://github.com/caridy/redefine-custom-elements), which does some strategic patches to HTMLElement and the Custom Element Registry – in a style similar to the scoped custom element polyfill – in order to get runtime control of control custom element registration and instantiation. With redefine-custom-elements active, re-registering a custom element tagname will result in the new constructor being used whenever that element is instantiated.
+
+On top of that, we're adding a lightweight protocol for custom element classes to participate in HMR. If a custom element class has a static method named `notifyOnHotReplace`, then when that custom element is re-registered, we will call that method on the original version of the class. The method has the signature:
+
+```typescript
+static notifyOnHotReplace(tagname: string, updatedClass: typeof this): void;
+```
+
+This is most of the information available to our patched custom element renderer, but fortunately it's all that's needed for a LitElement subclass to patch itself, update styles, then find all existing instances in the page and ask them to rerender.
+
+## Implementation Considerations
+
+### Implementation Plan
+
+The initial implementation will just be a single PR. We've been using something similar inside google3 for a little while now, and it seems to work decently well – though the HMR model is somewhat different in Google's proprietary stack.
+
+### Backward Compatibility
+
+This is an experimental API. We can remove the notifyOnHotReplace method from LitElement without breaking anything, because it's only present in dev mode, and it's typed as potentially undefined, so any user should expect that it may not be present.
+
+If we change the HMR element protocol in a non-backwards compatible way, we should rename the static method.
+
+### Testing Plan
+
+To test this properly, we will need integration tests with a real web dev server, using a browser-controlling system like Playwright. The initial PR may not have tests.
```

Comment:
"The initial PR may not have tests." - I wouldn't worry about saying that. Just saying that we'll have integration tests is fine.

Reply:
Done